File: integrations/sqlite/typescript/src/approval-format.ts
import type { DatabaseSync } from "node:sqlite";
import { StorageSession } from "purra";
import { OutputJournal } from "./journal.js";

export function storageVersion(db: DatabaseSync): 4 | 5 {
  if (!db.prepare("SELECT 1 FROM sqlite_master WHERE type='table' AND name='purra_state'").get()) return 4;
  const marker = db.prepare("SELECT version,body FROM purra_state WHERE scope='' AND sdk='purra.approvals'").get();
  const version = marker?.version === 5 && marker.body === "{}" ? 5 : 4;
  if (marker && version !== 5 || db.prepare("SELECT 1 FROM purra_state WHERE version != ? LIMIT 1").get(version)) throw new Error("unsupported SQLite storage version");
  return version;
}

export interface ApprovalUpgradeInspection {
  readonly schemaVersion: 1;
  readonly authority: "diagnosis_only";
  readonly storageVersion: 4 | 5;
  readonly targetVersion: 5;
  readonly status: "ready" | "blocked" | "already_enabled";
  readonly blockers: readonly string[];
}
function activationBlocker(db: DatabaseSync): string | undefined {
  for (const row of db.prepare("SELECT scope,sdk,body FROM purra_state ORDER BY scope,sdk").all()) {
    if (row.sdk !== "typescript") return "approval_activation_foreign_sdk";
    const journal = new OutputJournal(db, String(row.scope), false);
    const session = new StorageSession(String(row.body), "all", journal.restore());
    if (session.hasUnsettledExecution()) return "approval_activation_execution_pending";
  }
  return undefined;
}
export function inspectApprovalUpgrade(db: DatabaseSync): ApprovalUpgradeInspection {
  const version = storageVersion(db);
  const blocker = version === 5 ? undefined : activationBlocker(db);
  return Object.freeze({schemaVersion: 1, authority: "diagnosis_only", storageVersion: version,
    targetVersion: 5, status: version === 5 ? "already_enabled" : blocker ? "blocked" : "ready",
    blockers: Object.freeze(blocker ? [blocker] : [])});
}
export function enableApprovals(db: DatabaseSync): void {
  if (storageVersion(db) === 5) return;
  const blocker = activationBlocker(db);
  if (blocker) throw new Error(blocker);
  db.exec(`UPDATE purra_state SET version=5;
    INSERT INTO purra_state VALUES('', 'purra.approvals', 5, '{}');
    CREATE TABLE purra_approvals (
      scope TEXT NOT NULL, sdk TEXT NOT NULL, approval_id TEXT NOT NULL,
      run_id TEXT NOT NULL, call_id TEXT NOT NULL, body TEXT NOT NULL,
      PRIMARY KEY(scope,sdk,approval_id), UNIQUE(scope,sdk,run_id,call_id));
    CREATE TRIGGER purra_approval_format_insert BEFORE INSERT ON purra_state WHEN NEW.version != 5
      BEGIN SELECT RAISE(ABORT, 'unsupported SQLite storage version'); END;
    CREATE TRIGGER purra_approval_format_update BEFORE UPDATE ON purra_state WHEN NEW.version != 5
      BEGIN SELECT RAISE(ABORT, 'unsupported SQLite storage version'); END;`);
}

File: integrations/sqlite/python/src/purra_sqlite/approval_format.py
"""Opt-in database-wide v5 fence; legacy-only databases stay v4."""
from purra.storage import StorageSession
from .journal import OutputJournal


MARKER_SDK = "purra.approvals"


def storage_version(db):
    if not db.execute("SELECT 1 FROM sqlite_master WHERE type='table' AND name='purra_state'").fetchone():
        return 4
    marker = db.execute("SELECT version,body FROM purra_state WHERE scope='' AND sdk=?", (MARKER_SDK,)).fetchone()
    version = 5 if marker == (5, '{}') else 4
    if marker is not None and version != 5:
        raise ValueError("unsupported SQLite storage version")
    if db.execute("SELECT 1 FROM purra_state WHERE version != ? LIMIT 1", (version,)).fetchone():
        raise ValueError("unsupported SQLite storage version")
    return version


def _activation_blocker(db):
    for scope, sdk, body in db.execute("SELECT scope,sdk,body FROM purra_state ORDER BY scope,sdk").fetchall():
        if sdk != "python":
            return "approval_activation_foreign_sdk"
        session = StorageSession(body)
        OutputJournal(db, scope, initialize=False).restore(session)
        if session.has_unsettled_execution():
            return "approval_activation_execution_pending"
    return None


def inspect_approval_upgrade(db):
    """Observe upgrade readiness in the caller's read transaction; never authorize it."""
    version = storage_version(db)
    blocker = None if version == 5 else _activation_blocker(db)
    return {"schemaVersion": 1, "authority": "diagnosis_only", "storageVersion": version,
            "targetVersion": 5, "status": "already_enabled" if version == 5 else "blocked" if blocker else "ready",
            "blockers": [] if blocker is None else [blocker]}


async def enable_approvals(storage):
    async with storage._connection():
        if storage_version(storage._db) == 5:
            return
        blocker = _activation_blocker(storage._db)
        if blocker:
            raise ValueError(blocker)
        storage._db.execute("UPDATE purra_state SET version=5")
        storage._db.execute("INSERT INTO purra_state VALUES('', ?, 5, '{}')", (MARKER_SDK,))
        storage._db.execute("""CREATE TABLE purra_approvals (
            scope TEXT NOT NULL, sdk TEXT NOT NULL, approval_id TEXT NOT NULL,
            run_id TEXT NOT NULL, call_id TEXT NOT NULL, body TEXT NOT NULL,
            PRIMARY KEY(scope,sdk,approval_id), UNIQUE(scope,sdk,run_id,call_id))""")
        for operation in ("INSERT", "UPDATE"):
            storage._db.execute(f"""CREATE TRIGGER purra_approval_format_{operation.lower()}
                BEFORE {operation} ON purra_state WHEN NEW.version != 5
                BEGIN SELECT RAISE(ABORT, 'unsupported SQLite storage version'); END""")
